### ISSUE TITLE
Add protobuf support for deserializing messages

### ISSUE THREAD

**Issue by weeco** [cloudhut/kowl#124]
Users have reported via Email they are using Protobuf. It's been reported that the proto files are managed using a Git repository that contains all protos. It's also possible to use Confluent's schema registry to manage the proto files. Both should be supported. Currently it's probably more common that protos are managed in a Git repository as the support for Protobuf in the schema registry is a rather new feature.

**Challenges**

As of now we automatically detect the required deserializer. With protobuf there's no such thing like a magic byte which indicates that the given messages is serialized with Protobuf. Therefore we must offer users an option to choose the deserialization method and furthermore what proto the backend should use to deserialize a given message (key/value).

**Comment by flolu** [comment]
I would really appreciate this feature. :tada: :+1: 

But how would you know which Protobuf message belongs to which incoming message?

I think the user would need to send along an `id` in the message header and provide a mapping between the `id` and the respective Protobuf message declaration.

**Comment by weeco** [comment]
Currently we think manual user input is required at least once to create some sort of mapping between the protofile and the messages.

We could do more advanced stuff like trying to decode a given message with all available protos and if any of the decoded results is either a valid JSON (1st) or valid ASCII string (2nd) we could suggest to use this proto for a given topic.

I think the schema registry nowadays supports Protobuf and I assume there's some mapping between the topic name and the respective protobuf as well. I'd need to take a look at this

**Comment by BnMcG** [comment]
I'd  be happy to have a go at contributing towards this if anybody can give me any pointers for where to start implementing? I have some data and some proto definitions to test with.

**Comment by weeco** [comment]
I believe this issue is a rather big task, which is very vaguely defined. Some desired features we want to achieve:
- Support Protobuf regardless whether schema registry, git repository or mounted files/directory are used for the proto definitions
- To make it as comfortable as possible Kowl shall guess the right proto definition for a given topic/message. To do that we want take one message and try to deserialize this message with all available proto definitions. The first definition that deserializes the message to an ASCII string (maybe there are more accurate ways of assuming correct deserializations) will be suggested to be used in the frontend.
- Protobuf serialized messages can also be used in the JavaScript filters and hence the deserialized messages will be presented/converted as JSON. We do that with Avro and XML already

However I think this can issue can be divided in some subtasks which could be a good start for first time contributors. Some subtasks:

1. We need to add a config options for users that tell Kowl where (and how) to look for proto files (Git/Schema registry/Mounted directory). Config should be added here: https://github.com/cloudhut/kowl/blob/master/backend/pkg/kafka/config.go . If it's git we can probably reuse the Git client/serivce we built that fetched topic documentation already (https://github.com/cloudhut/kowl/tree/master/backend/pkg/git) . All these proto definitions must be loaded and made available to the deserializer which is initialized here: https://github.com/cloudhut/kowl/blob/master/backend/pkg/kafka/service.go#L74
2. We need some code for the deserializer that deserializes a protobuf serialized `[]byte` (key/value) which then has to be serialized into JSON at some point. That's what we already do for avro, xml etc.: https://github.com/cloudhut/kowl/blob/master/backend/pkg/kafka/deserializer.go
3. The `deserializePayload(payload []byte)` method currently tries all available deserialization options for every single message. That was okay because this was a rather lightweight process. However with our new guessing approach we can't do this for every message anymore, as this might be called thousands of times per second (for streaming search with filters). Therefore we need to add some hint if protobuf is used that tells us "we want to use protobuf and please use that proto to deserialize". I'm not yet sure where this is hint is supposed to come from (is it the initial guess from the backend? or is a user required to manually select a proto file in the frontend for a given topic first?).

Feel free to tackle some of these issues and we can hop on the PR and possibly help or later on finish it. You are invited to join our Discord server as well (see readme badge), which would make it a bit easier to collobarate on this.

**Comment by BnMcG** [comment]
I was going to start with the use case that the protobuf spec was added to a Confluent schema registry as that seemed to be the quickest way to get off the ground in terms of actually getting the protobuf schema into Kowl - I took a look at it on Sunday evening though and wasn't too sure how to add a Protobuf schema via the registry API, it looks like support for it is new and the documented ways I've seen use their Java classes to automatically add the schema to the registry if it's not already there.

I might continue having a look at that tonight and join your Discord server, too. Once I work out how to add a schema to the registry via the API it seemed fairly straight forward to read it into Kowl (based on the existing Avro code), and from there the Protobuf deserialization can be done, leaving only reading schemas from a repository later. 

I'm also not too sure on how to do the deserialization, after a cursory glance on Sunday it seems like we'll have to have an object that is capable of parsing a Protobuf schema to determine the field types and then read the `[]byte` slice to convert to a string, I'm not sure if there's a library already out there that can parse the schema for us, but I've not found anything like it yet, since of course Protobuf is usually used with classes compiled from the spec.

**Comment by clarknova9** [comment]
Support for deserializing protobuf messages would be awesome. Lack of protobuf support is the only thing preventing me from rolling Kowl out in my org. 

It would be great if protobuf deserialization could be configured similarly to how it's done on akhq. 
```
      deserialization:
        protobuf:
          protobuf-folder: "/app/proto"
          topics-mapping:
            - topic-name: "topic.name"
              proto-file: "message.proto"
              message-type: "Typename"
```

Mount a folder with the proto files and map each topic to the appropriate proto file and message type. 

@weeco I saw your comment about having Kowl try each protobuf definition until it finds the right one. Honestly, unless it can be done with minimal hit to performance, I wouldn't manually mapping each topic to its definition in the config. 

Fantastic tool all around. I've tried all the Kafka tools out there. Kowl is by far my favorite. Splendid work!

**Comment by weeco** [comment]
Thanks for the feedback and providing the example config @clarknova9 .

> Honestly, unless it can be done with minimal hit to performance, I wouldn't manually mapping each topic to its definition in the config.

Of course the performance is critical, but I think this should be no problem. Let's assume you have 1000 different proto messages. When we consume a topic we could sort all proto message type names by similarity to the topic name. Then we could try deserializing a given message in that order for a single message (or actually all in parallel - shouldn't matter either). All of this should be done within a few milliseconds. At least that's the theory, we haven't tried it in practice yet.

Given the user feedback Protobuf support seems to be important (and also a showstopper if there's no support) for many people. Thus I'll be working on this as soon as possible 👍

**Comment by weeco** [comment]
@clarknova9 @BnMcG @flolu I just merged Protobuf support into master and also added documentation along with it that describes how it is supposed to be configured: https://github.com/cloudhut/kowl/blob/master/docs/features/protobuf.md

I don't have access to a real world production environment that uses Protobuf, so further feedback is highly appreciated.

**Comment by feldblyum** [comment]
> @clarknova9 @BnMcG @flolu I just merged Protobuf support into master and also added documentation along with it that describes how it is supposed to be configured: https://github.com/cloudhut/kowl/blob/master/docs/features/protobuf.md
> 
> I don't have access to a real world production environment that uses Protobuf, so further feedback is highly appreciated.

@weeco Excited to test this out! Unfortunately, when I run the docker image **kowl:master** I'm unable to connect to my brokers. I didn't have this problem with **kowl:1.2.1** or **kowl:1.2.2**. A 500 error is returned on the /topics page. I don't know if this matter, but I'm connecting to an AWS Kafka Cluster (MSK). 

Let me know if you would like me to create a new issue for this since this one is closed. 


```
{"level":"info","ts":"2021-01-10T21:29:01.763Z","msg":"started Kowl","version":"master","git_sha":"6040e5dd007b727557fe30d8b5b03d16935b72c0","built":"2021-01-10T20:50:57Z"}
{"level":"info","ts":"2021-01-10T21:29:01.786Z","msg":"connecting to Kafka seed brokers, trying to fetch cluster metadata"}
{"level":"debug","ts":"2021-01-10T21:29:01.786Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.296412185}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:02.151Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:02.151Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:02.218Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:02.218Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"info","ts":"2021-01-10T21:29:02.288Z","msg":"successfully connected to kafka cluster","advertised_broker_count":12,"topic_count":33,"controller_id":2}
{"level":"debug","ts":"2021-01-10T21:29:02.288Z","msg":"using compression for all http routes","level":4}
{"level":"info","ts":"2021-01-10T21:29:02.290Z","msg":"Server listening on address","address":"[::]:8080","port":8080}
{"level":"debug","ts":"2021-01-10T21:29:33.907Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.381542293}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:34.361Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:34.361Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:34.434Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:34.434Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.507Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.337782025}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:34.918Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:34.918Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:34.991Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:34.991Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:35.064Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.313360247}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:35.444Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:35.444Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:35.511Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:35.511Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.578Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.34766203}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:35.995Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:35.995Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:36.063Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:36.063Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"config response resource has an error","resource_name":"test-topic","error":"TOPIC_AUTHORIZATION_FAILED"}
{"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"Sending REST error","route":"/api/topics","method":"GET","status_code":500,"remote_address":"10.113.0.26:36496","public_error":"Could not list topics from Kafka cluster","error":"TOPIC_AUTHORIZATION_FAILED"}

```

**Comment by clarknova9** [comment]
> @clarknova9 @BnMcG @flolu I just merged Protobuf support into master and also added documentation along with it that describes how it is supposed to be configured: https://github.com/cloudhut/kowl/blob/master/docs/features/protobuf.md
> 
> I don't have access to a real world production environment that uses Protobuf, so further feedback is highly appreciated.

@weeco Excited to test this out! Unfortunately, when I run the docker image kowl:master I'm unable to connect to my brokers. I didn't have this problem with kowl:1.2.1 or kowl:1.2.2. A 500 error is returned on the /topics page. I don't know if this matter, but I'm connecting to an AWS Kafka Cluster (MSK).

Let me know if you would like me to create a new issue for this since this one is closed.
```
{"level":"info","ts":"2021-01-10T21:29:01.763Z","msg":"started Kowl","version":"master","git_sha":"6040e5dd007b727557fe30d8b5b03d16935b72c0","built":"2021-01-10T20:50:57Z"}
{"level":"info","ts":"2021-01-10T21:29:01.786Z","msg":"connecting to Kafka seed brokers, trying to fetch cluster metadata"}
{"level":"debug","ts":"2021-01-10T21:29:01.786Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.296412185}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"debug","ts":"2021-01-10T21:29:02.083Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:02.151Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:02.151Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:02.218Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:02.218Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-1.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":-2147483648}
{"level":"info","ts":"2021-01-10T21:29:02.288Z","msg":"successfully connected to kafka cluster","advertised_broker_count":12,"topic_count":33,"controller_id":2}
{"level":"debug","ts":"2021-01-10T21:29:02.288Z","msg":"using compression for all http routes","level":4}
{"level":"info","ts":"2021-01-10T21:29:02.290Z","msg":"Server listening on address","address":"[::]:8080","port":8080}
{"level":"debug","ts":"2021-01-10T21:29:33.907Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.381542293}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.289Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:34.361Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:34.361Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:34.434Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:34.434Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-10.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":10}
{"level":"debug","ts":"2021-01-10T21:29:34.507Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.337782025}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:34.845Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:34.918Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:34.918Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:34.991Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:34.991Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-2.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":2}
{"level":"debug","ts":"2021-01-10T21:29:35.064Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.313360247}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.377Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:35.444Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:35.444Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:35.511Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:35.511Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-8.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":8}
{"level":"debug","ts":"2021-01-10T21:29:35.578Z","msg":"opening connection to broker","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"kafka connection succeeded","source":"kafka_client_hooks","host":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com","dial_duration":0.34766203}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"connection opened to broker","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"debug","ts":"2021-01-10T21:29:35.926Z","msg":"issuing api versions request","source":"kafka_client","version":3}
{"level":"debug","ts":"2021-01-10T21:29:35.995Z","msg":"kafka does not know our ApiVersions version, downgrading to version 0 and retrying","source":"kafka_client"}
{"level":"debug","ts":"2021-01-10T21:29:35.995Z","msg":"issuing api versions request","source":"kafka_client","version":0}
{"level":"debug","ts":"2021-01-10T21:29:36.063Z","msg":"initialized api versions","source":"kafka_client","versions":[7,10,5,7,2,1,5,2,6,5,2,4,2,2,2,2,2,1,2,3,3,1,1,2,1,1,1,0,2,1,1,1,2,1,1,1,1,1,1,1,1,1,1,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]}
{"level":"debug","ts":"2021-01-10T21:29:36.063Z","msg":"connection initialized successfully","source":"kafka_client","addr":"b-12.test-v2-kafka-mtls.c2.kafka.eu-west-1.amazonaws.com:9094","id":12}
{"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"config response resource has an error","resource_name":"test-topic","error":"TOPIC_AUTHORIZATION_FAILED"}
{"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"Sending REST error","route":"/api/topics","method":"GET","status_code":500,"remote_address":"10.113.0.26:36496","public_error":"Could not list topics from Kafka cluster","error":"TOPIC_AUTHORIZATION_FAILED"}
```

**Comment by weeco** [comment]
@clarknova9 THanks for the quick feedback. This is likely caused by our Kafka library change. We switched from sarama to a new library that offers a lot more of Kafka features (especially the newer KIPs!). I just upgraded the library where a more detailed error message is returned. My initial impression though is that your user doesn't have the permission to describe the topic's config

> {"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"config response resource has an error","resource_name":"test-topic","error":"TOPIC_AUTHORIZATION_FAILED"}

From the Kafka docs:


> // TOPIC_AUTHORIZATION_FAILED is returned if asking to describe topic
> // configs but the client is not authorized to do so.

**Comment by clarknova9** [comment]
> @clarknova9 THanks for the quick feedback. This is likely caused by our Kafka library change. We switched from sarama to a new library that offers a lot more of Kafka features (especially the newer KIPs!). I just upgraded the library where a more detailed error message is returned (a new docker image is building right now). My initial impression though is that your user doesn't have the permission to describe the topic's config
> 
> > {"level":"error","ts":"2021-01-10T21:29:36.135Z","msg":"config response resource has an error","resource_name":"test-topic","error":"TOPIC_AUTHORIZATION_FAILED"}
> 
> From the Kafka docs:
> 
> > // TOPIC_AUTHORIZATION_FAILED is returned if asking to describe topic
> > // configs but the client is not authorized to do so.

@weeco Ah, my account probably does not have that permission. I don't know exactly which permissions were given to my user account, but it is meant to be consumer only. It should have access to list all topics and consume from all topics. Will that be a problem?

**Comment by weeco** [comment]
> @weeco Ah, my account probably does not have that permission. I don't know exactly which permissions were given to my user account, but it is meant to be consumer only. It should have access to list all topics and consume from all topics. Will that be a problem?

It will require a few more permissions for all the features Kowl provides. For instance you need to be able to describe topic configs, broker configs, describe log dir sizes etc.. I don't know the exact ACL permission set to cover all functionalities, but we could cover that in the documentation at some point.

**Comment by clarknova9** [comment]
> > @weeco Ah, my account probably does not have that permission. I don't know exactly which permissions were given to my user account, but it is meant to be consumer only. It should have access to list all topics and consume from all topics. Will that be a problem?
> 
> It will require a few more permissions for all the features Kowl provides. For instance you need to be able to describe topic configs, broker configs, describe log dir sizes list all groups, get group offsets, etc.. I don't know the exact ACL permission set to cover all functionalities, but we could cover that in the documentation at some point.

@weeco Is there anything in the configuration settings now that would allow kowl to load the features available per the user permissions and only throw the errors when a specific feature  does not have the necessary permissions?

**Comment by weeco** [comment]
> @weeco Is there anything in the configuration settings now that would allow kowl to load the features available per the user permissions and only throw the errors when a specific feature does not have the necessary permissions?

I agree it should behave like this (basically grey out features, that couldn't be fetched in the backend), but we still have to get to that in the course of time. Currently this is not possible.

**Comment by dev501-code** [comment]
@weeco 
My proto files are in multiple git repos. Can we support pull from multiple git repos? Also, instead of a git repo, can we export/copy all .proto files into a particular folder and kowl picks up the proto files from that folder?

**Comment by weeco** [comment]
@dev501-code Sounds reasonable to add another provider (local filesystem) for providing proto files. Do you mind filing a separate issue for that?

**Comment by atoom** [comment]
@weeco Would it be possible to automatically fetch the proto files from the schema registry - since proto is supported there now as well ?

**Comment by weeco** [comment]
@atoom Currently not supported, we assume most wouldn't need that as of now since this is a pretty new feature. I'd also recommend submitting a separate issue for that. In theory of course possible to add 👍
